pkg/trace/writer/trace.go

```diff
@@ -71,20 +85,21 @@ type TraceWriter struct {
 
 // NewTraceWriter returns a new TraceWriter. It is created for the given agent configuration and
 // will accept incoming spans via the in channel.
-func NewTraceWriter(cfg *config.AgentConfig) *TraceWriter {
+func NewTraceWriter(cfg *config.AgentConfig, prioritySampler prioritySampler, errorsSampler errorsSampler, rareSampler rareSampler) *TraceWriter {
```

Comment:
I am personally not a fan of this dependency but this was most obvious way to solve the issue without having duplicated states. Another similar idea would have been for the writer to access an `AgentPayloadBuilder` so the logic could come from `agent.go`. Anyway please chime in if you have an opinion, happy to make the change.